[Tooling] Ruby-fy `send_strings_for_translation` & `commit_strings` lanes

AliSoftware opened the pull request:
### Description

I noticed that the `send_strings_for_translation` and `commit_strings` lanes were written by calling `sh` directly to run some `mkdir`+`cp`+`git add` + `git commit` commands.

 - This PR replaces it with more idiomatic approach using Ruby code (`FileUtils.mkdir_p`, `FileUtils.cp`, fastlane's `git_add` and `git_commit` actions) for readability and consistency.
 - It also stops making those lanes do the `git push`, since the only caller of those helper lanes—which is `complete_code_freeze` lane—ends up already calling `push_to_git_remote` after calling those two.
 - It also now enables `auto_commit:true` by default on `commit_strings` lane, while I noticed that it was not on by default when called from `complete_code_freeze`->`localize_libs`->`commit_strings`, which could have led to CI job failing because CI is not interactive, while `complete_code_freeze` is now run from CI and not from RM's Macs anymore
 - Finally I renamed `send_strings_for_translation` to `freeze_strings_for_translation` so its name would be more accurate, and added some comments about how the WPCOM job is pulling those into GlotPress.

### Testing

 - Cut a testing branch from this PR's head branch, `git push -u` it (tracking it so it'd be easier to delete it from remote alongside the local branch at the end of the test)
 - Testing `commit_strings`:
    - Make dummy changes to the `values/strings.xml` file
    - Change `private_lane` to `lane` on `commit_strings`
    - Run `bundle exec fastlane commit_strings auto_commit:false` and verify it prints the prompt message. Type enter (without doing anything) to validate the prompt and exit the lane
    - Run `bundle exec fastlane commit_strings` and verify it created a commit containing your dummy changes from `values/strings.xml`
 - Testing `freeze_strings_for_translation`:
    - `git rm -r fastlane/resources/values`, then commit the deletion
    - Change `private_lane` to `lane` on `freeze_strings_for_translation`
    - Run `bundle exec fastlane freeze_strings_for_translation` and verify it re-created the `fastlane/resources/values/` folder and created a commit containing the `fastlane/resources/values/strings.xml` file
 - Delete your testing branch from git remote.

### Target branch

Targeting `release/19.2` because it already contains changes to the `send_strings_for_translations` lane when Thomaz fixed the missing `/`, so targeting `trunk` would have led to conflicts. Will probably rebase to target `trunk` once the `release/19.2->trunk` backmerge PR post-code-freeze have landed though.

AliSoftware committed:
Ruby-fy `send_strings_for_translation` lane

AliSoftware committed:
Ruby-fy `commit_strings` lane

AliSoftware committed:
Don't make helper lanes git-push

As the caller of those lanes, `complete_code_freeze`, already includes a `push_to_git_remote` call after calling those two lanes. And they are not called from anywhere else.

Use the opportunity to rename `send_strings_for_translation` to `freeze_strings_for_translation` which seemed more appropriate, as we're not the ones sending the strings to GlotPress but instead are just updating the frozen strings file that gets pulled by the WPCOM job.

AliSoftware committed:
Move lane, add comment, set `auto_commit: true`

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iangmaia commented on fastlane/Fastfile:
```diff
@@ -41,7 +41,7 @@ DEV_RELEASE_NOTES_PATH = {
 RAW_SCREENSHOTS_PROCESSING_DIR = File.join(Dir.pwd, 'screenshots', 'raw_tmp')
 PROMO_SCREENSHOTS_DIR = File.join(Dir.pwd, 'screenshots', 'promo_screenshots')
-FROZEN_STRINGS_DIR_PATH = File.join(Dir.pwd, 'resources', 'values')
+FROZEN_STRINGS_PATH = File.join(Dir.pwd, 'resources', 'values', 'strings.xml')
```
👍

iangmaia commented on fastlane/Fastfile:
```diff
@@ -1152,10 +1143,45 @@ platform :android do
+  private_lane :commit_strings do |auto_commit: true|
+    if auto_commit
+      git_add(path: MAIN_STRINGS_PATH)
+      git_commit(
```
While right after calling `freeze_strings_for_translation` we also `push_to_git_remote(tags: false)`, when calling `commit_strings` we don't really do so, and the old lane code called `sh('git push origin HEAD')`.

I guess we could also follow the same pattern and call `push_to_git_remote(tags: false)` right after `commit_strings(auto_commit: options[:auto_commit]) unless is_repo_clean`?

AliSoftware replied:
> when calling `commit_strings` we don't really do so
Don't we though?

From a quick ⌘F in the Fastfile, `commit_strings` is only called from the `localize_libs` lane, and that `localize_libs` lane is only called by `complete_code_freeze` (actually, right before calling `freeze_strings_for_translation`)… so I think we still have a guarantee that the `push_to_git_remote(tags: false)` from line 202 will push both the changes from `commit_strings` and from `freeze_strings_for_translations`, right?

iangmaia replied:
Ah, good point, I was just checking the diff and seeing that previously this lane already pushed (while now it doesn't), but indeed `push_to_git_remote(tags: false)` will be done later in `complete_code_freeze`, after `send_strings_for_translation` 👍

iangmaia commented on fastlane/Fastfile:
```diff
@@ -754,7 +754,7 @@ platform :android do
     end
 
     is_repo_clean = `git status --porcelain`.empty?
```
While at it, we could also call the more standard `ensure_git_status_clean`?

AliSoftware replied:
I thought it was a good idea at first, but I just realized that `ensure_git_status_clean` will expect the git status to be clean… and [will `UI.user_error` if not](https://github.com/fastlane/fastlane/blob/5aa842986cbaf5298623e340241623d73901f35f/fastlane/lib/fastlane/actions/ensure_git_status_clean.rb#L47).

While here what we want is just to test and have a yes/no answer… so that wouldn't fit the use case after all 😒

iangmaia replied:
Ah indeed -- if it's not clean we don't want it to fail and this is still part of the happy flow.

AliSoftware replied:
To be honest, now that WCAndroid uses Release-on-CI, that means that `auto_commit` will probably always be used with `true`, which means that `commit_strings` will end up only being a call to `git_add`+`git_commit`… and we can always call those 2 actions unconditionally (without having to manually test ourselves if repo is dirty or not first) as long as we set `allow_nothing_to_commit: true` on the `git_commit` call…

So maybe we should remove the `auto_commit:` parameter and the `is_repo_clean` test and just do that… at which point we'd probably not need a separate `commit_strings` lane anymore, and can move the `git_add` + `git_commit` directly inside the `localize_libs` lane…

AliSoftware replied:
> So maybe we should remove the auto_commit: parameter and the is_repo_clean test and just do that… at which point we'd probably not need a separate commit_strings lane anymore, and can move the git_add + git_commit directly inside the localize_libs lane…

Implemented in e993e0236491d8c927f7a9b0da36ddfd2121e177

Tested by removing 3 arbitrary `<string name="about_automattic_*">` entries from `WooCommerce/src/main/res/values/strings.xml`, committing the removal, then running `bundle exec fastlane localize_libs` and confirm that it created a new commit adding those 3 entries back.

AliSoftware commented:
Rebased on `trunk` now that the codefreeze backmerge PR has landed, so this is now officially Ready for review 🙂

iangmaia commented on fastlane/Fastfile:
```diff
@@ -1152,10 +1143,45 @@ platform :android do
+    git_add(path: FROZEN_STRINGS_PATH)
+    git_commit(
+      path: FROZEN_STRINGS_PATH,
+      message: 'Send strings to translation.',
```
It was already like that but I noticed also due to the difference between the method name and the commit message.

```suggestion
      message: 'Freeze strings for translation.',
```

iangmaia reviewed (approved):
Looks good. Also tested it and everything worked well 👍

AliSoftware committed:
Adjust commit message wording

Co-authored-by: Ian Guedes Maia <iangmaia@users.noreply.github.com>

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/223">19.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AliSoftware committed:
Remove `commit_strings` lane and inline its code

 - It was only used in `localize_libs` and only made sense for that lane anyway
 - Now that this app uses Release-on-CI, we'll never really need `auto_commit: false` anymore (it was there from the time when we ran those lanes locally on our Macs and wanted to double-check the changes before committing while the implementation was fragile; but this is obsolete now)

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -753,8 +753,11 @@ platform :android do
+    git_commit(
+      path: MAIN_STRINGS_PATH,
+      message: 'Include lib strings for translation',
+      allow_nothing_to_commit: true
+    )
```
Note that:
 - It's only necessary to call `git_add` explicitly before calling `git_commit` if that file is not in git yet (or might have been deleted since). Since `MAIN_STRINGS_PATH` is guaranteed to exist already in git, it's thus unnecessary to call `git_add` in this particular case (while for files we delete (like changelogs) or that we regenerate, it's still useful to)
 - We don't need to test `is_repo_clean` anymore because `allow_nothing_to_commit: true` already takes care of checking this (and being a no-op not creating any commit if there's no changes)

AliSoftware committed:
Fix rubocop violations, update YARD
